pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -183,11 +190,22 @@ public synchronized void channelInactive(ChannelHandlerContext ctx) throws Excep
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-        state = State.Closing;
         super.exceptionCaught(ctx, cause);
-        LOG.warn("[{}] Got exception {} : {} {}", remoteAddress, cause.getClass().getSimpleName(), cause.getMessage(),
+        LOG.warn("[{}] Got exception {} : Message: {} State: {}", remoteAddress, cause.getClass().getSimpleName(),
+                cause.getMessage(), state,
                 ClientCnx.isKnownException(cause) ? null : cause);
-        ctx.close();
+        if (state != State.Closed) {
+            state = State.Closing;
+        }
+        if (ctx.channel().isOpen()) {
```

Comment:
why can't we blindly call "ctx.close()" ?

Reply:
it could be called, but it it's already closed, it might not close the directProxyHandler (if that happens to be present). The else branch is the reason for adding this logic.

Reply:
If it is already closed, why wouldn't `directProxyHandler` also be closed? Isn't the `channelInactive` method always called when the channel is closed?

Reply:
I see that the `directProxyHandler` might not be closed yet, but wouldn't we expect it to be closed soon, so we don't need to close it in this method?